packages/bitcore-lib/lib/script/interpreter.js

```diff
@@ -70,26 +72,133 @@ Interpreter.prototype.verifyWitnessProgram = function(version, program, witness,
+    if (!(flags & Interpreter.SCRIPT_VERIFY_TAPROOT)) {
+      return true;
+    }
+    stack = Array.from(witness);
+    if (stack.length == 0) {
+      this.errstr = 'SCRIPT_ERR_WITNESS_PROGRAM_WITNESS_EMPTY';
+      return false;
+    }
+    if (stack.length >= 2 && stack[stack.length - 1].length && stack[stack.length - 1][0] === Script.ANNEX_TAG) {
+      // Drop annex (this is non-standard; see IsWitnessStandard)
+      const annex = stack.pop();
+      const annexWriter = new BufferWriter();
+      annexWriter.writeVarintNum(annex.length);
+      annexWriter.write(annex);
+      execdata.annexHash = Hash.sha256(annexWriter.toBuffer());
+      execdata.annexPresent = true;
+    }
+    execdata.annexInit = true;
+    if (stack.length === 1) {
+      // Key path spending (stack size is 1 after removing optional annex)
+      if (!this.checkSchnorrSignature(stack[0], program, Signature.Version.TAPROOT, execdata)) {
+        return false;
+      }
+      return true;
+    } else {
+      // Script path spending (stack size is >1 after removing optional annex)
+      const control = stack.pop();
+      const scriptPubKeyBuf = stack.pop();
+
+      if (
+        control.length < Interpreter.TAPROOT_CONTROL_BASE_SIZE  ||
+        control.length > Interpreter.TAPROOT_CONTROL_MAX_SIZE   ||
+        ((control.length - Interpreter.TAPROOT_CONTROL_BASE_SIZE) % Interpreter.TAPROOT_CONTROL_NODE_SIZE) != 0
+      ) {
+        this.errstr = 'SCRIPT_ERR_TAPROOT_WRONG_CONTROL_SIZE';
+        return false;
+      }
+      // TODO
+      execdata.tapleafHash = Interpreter.computeTapleafHash(control[0] & Interpreter.TAPROOT_LEAF_MASK, scriptPubKeyBuf);
+      if (!Interpreter.verifyTaprootCommitment(control, program, execdata.tapleafHash)) {
+        this.errstr = 'SCRIPT_ERR_WITNESS_PROGRAM_MISMATCH';
+        return false;
+      }
+      execdata.tapleafHashInit = true;
+      if ((control[0] & Interpreter.TAPROOT_LEAF_MASK) === Interpreter.TAPROOT_LEAF_TAPSCRIPT) {
+        // Tapscript (leaf version 0xc0)
+        let witnessSize;
+        {
+          const bw = new BufferWriter();
+          bw.writeVarintNum(witness.length);
+          for (let element of witness) {
+            bw.writeVarintNum(element.length);
+            bw.write(element);
+          }
+          witnessSize = bw.toBuffer().length;
+        }
+
+        try {
+          scriptPubKey = new Script(scriptPubKeyBuf);
+        } catch (err) {
+          // Note how this condition would not be reached if an unknown OP_SUCCESSx was found
+          this.errstr = 'SCRIPT_ERR_BAD_OPCODE';
+          return false;
+        }
+
+        execdata.validationWeightLeft = witnessSize + Script.VALIDATION_WEIGHT_OFFSET;
+        execdata.validationWeightLeftInit = true;
+        return this.executeWitnessScript(scriptPubKey, stack, Signature.Version.TAPSCRIPT, satoshis, flags, execdata);
+      }
+      if (flags & Interpreter.SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_TAPROOT_VERSION) {
```

Comment:
Are there superceding cases above? Could there be? Or is this a condition which should supercede and return false?

If so, putting this simple check at the top could improve readability.